[FEATURE] PAGEVIEW - Optimized integration of Page Rendering via Fluid (#1235)
TYPO3-Documentation/TYPO3CMS-Reference-Typoscript: Documentation/ContentObjects/Pageview/Index.rst

..  include:: /Includes.rst.txt
..  index:: Content objects; PAGEVIEW
..  _cobj-pageview:

========
PAGEVIEW
========

..  versionadded:: 13.1

    This API is considered experimental until TYPO3 v13 LTS. It will be further
    adapted to be used for further rendering of
    custom Content Elements and Content Blocks.

The content object :typoscript:`EXTBASEPLUGIN` allows to render
:ref:`Extbase <t3coreapi:extbase>` plugins.

The new :typoscript:`PAGEVIEW` content object has very specific conventions and

> defaults, that requires (and allows) less configuration, with the benefit that

Reviewer: ```suggestion
defaults, that requires (and allows) less configuration. The benefit is that
```

following these conventions means less boilerplate code to maintain.

> If you follow these conventions, a few directories and files most follow the

Reviewer: ```suggestion
If you follow these conventions, a few directories and files must follow the
```

    path like :file:`EXT:my_extension/Resources/Private/Templates/MyPages`.

    The templates are expected in a subfolder :path:`Pages` or
    :path:`pages`.

    Fluid partials are looked up in a sub-directory called :path:`Partials` or
    :path:`partials`, layouts in :path:`Layouts` or :path:`layouts`.

    The name of the used page layout (:ref:`Backend layout <t3coreapi:be-layout>`)
    is resolved automatically.

    The paths are evaluated from highest to lowest priority.

..  _cobj-pageview-paths-example:

Example: Define a path that contains all templates
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This is a basic definition of a :fluid:`PAGEVIEW` object with only one path to the
templates:

..  code-block:: typoscript
    :caption: EXT:my_sitepackage/Configuration/TypoScript/setup.typoscript

    page = PAGE
    page.10 = PAGEVIEW
    page.10.paths.100 = EXT:my_sitepackage/Resources/Private/Templates/

The content of the folder could look like this:

..  directory-tree::
    :level: 2

    *   EXT:my_sitepackage/Resources/Private/Templates/

        *   Layouts

            *   Default.html
            *   WithoutHeader.html

        *   Pages

>             *   Default.html

Reviewer: Here we can add a hint, that the first letter has to be upper case. At least for now...

Author: I added a note

:file:`EXT:my_sitepackage/Resources/Private/Templates/Pages/TwoColumns.html`.

For all these templates
:ref:`partials <t3viewhelper:render_partial>`
are expected in folder
:path:`EXT:my_sitepackage/Resources/Private/Templates/Pages/Partials` and
:ref:`layouts <t3viewhelper:typo3fluid-fluid-layout>` in
:path:`EXT:my_sitepackage/Resources/Private/Templates/Pages/Layouts`.

..  _cobj-pageview-paths-example-extended:

Example: Define fallbacks for a template paths
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can use the directories defined in :confval:`pageview-paths` to
define fallback directories for the templates:

.. code-block:: typoscript

    page = PAGE
    page {
        10 = PAGEVIEW
        10.paths {
            100 = EXT:my_basic_sitepackage/Resources/Private/Templates/
            200 = EXT:my_general_sitepackage/Resources/Private/Templates/
            300 = EXT:my_special_sitepackage/Resources/Private/Templates/
        }
    }

The template for a page with a certain backend layout is first searched in
:path:`EXT:my_special_sitepackage/Resources/Private/Templates/Pages/` then in
:path:`EXT:my_general_sitepackage/Resources/Private/Templates/Pages/` and last

> in :path:`EXT:my_special_sitepackage/Resources/Private/Templates/Pages/`.

Reviewer: ```suggestion
in :path:`EXT:my_basic_sitepackage/Resources/Private/Templates/Pages/`.
```

..  _cobj-pageview-variables:

variables
---------

..  confval:: variables.[variable_name]
    :name: pageview-variables
    :type: :ref:`Content Object <cobject>`

    Sets variables that should be available in Fluid.

> Example: Make additional variables available in the Fluid template

Reviewer: Is the template a fluid template? I thought that it is called now "Pageview template"? But I don't know so much about this...

Author: The template is still in Fluid. Only the TypoScript object is now called "PAGEVIEW"

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

..  literalinclude:: _includes/_pageWithVariables.typoscript
    :langugage: typoscript
    :caption: EXT:my_sitepackage/Configuration/TypoScript/setup.typoscript

> The following variables are now available in the Fluid template:

Reviewer: Here the same... it isn't it a pageview template instead of a fluid template? Or is it a fluidtemplate just set up by a page template?

Author: "Fluid" is the markup language. "PAGEVIEW" is the name of the TypoScript object

..  include:: /Includes.rst.txt
..  index:: Content objects; PAGEVIEW
..  _cobj-pageview:

========
PAGEVIEW
========

..  versionadded:: 13.1

    This API is considered experimental until TYPO3 v13 LTS. It will be further
    adapted to be used for further rendering of

>     custom Content Elements and Content Blocks.

Reviewer 2: Do we already maybe have link targets? Should we link to ext:contenr-blocks for the time
being?

Author: I think this is a very temporary note that will be removed as soon as the PAGEVIEW is not experimental anymore?

    <f:section name="Main">
        <main role="main">
            <p>The site title in the current template is: {language.websiteTitle}</p>
        </main>
    </f:section>

.. _cobj-pageview-data-page:

page
----

..  confval:: page
    :name: pageview-data-page
    :type: :php:`TYPO3\CMS\Frontend\Page\PageInformation`

    The current `PageInformation` as object. See also
    :ref:`Frontend page information <t3coreapi:typo3-request-attribute-frontend-page-information>`.

.. _cobj-pageview-data-page-example:

Example: Display the title and abstract of the current page
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

..  code-block:: html
    :caption: EXT:my_sitepackage/Resources/Private/Templates/Pages/Default.html

    <f:layout name="Default" />
    <f:section name="Main">
        <main role="main">
            <p>The title of the page with id {page.id} is: {page.pageRecord.title}. It has the
                following abstract:</p>
            <p>{page.pageRecord.abstract}</p>
        </main>
    </f:section>

.. _cobj-pageview-data-settings:

settings
--------

..  confval:: settings
    :name: pageview-data-settings
    :type: array

>     Contains all TypoScript settings (= Constants)

Reviewer 2: Link to TypoScript Constants?

Author: I actually asked Stefan Bürk about this as I cannot really explain this variable yet

Author: I overhauled the section and added examples

                          <f:link.page pageUid="{menuItemLevel2.data.uid}"
                                       class="dropdown-item {f:if(condition: menuItemLevel2.active, then: 'active')}">
                             {menuItemLevel2.title}
                          </f:link.page>
                     </f:for>
                  </div>
               </f:then>
               <f:else>
                  <!-- Item has no children -->
                  <f:link.page pageUid="{menuItem.data.uid}" class="nav-link {f:if(condition: menuItem.active, then:'active')}">
                     {menuItem.title}
                  </f:link.page>
               </f:else>
            </f:if>
         </li>
      </f:for>
   </ul>

..  _cobj-pageview-paths:

paths
-----

..  confval:: paths.[priority]
    :name: pageview-paths
    :type: :ref:`data-type-path` with :ref:`stdWrap <stdwrap>`

    Sets an array of paths for the Fluid templates, usually
    :file:`EXT:my_extension/Resources/Private/Templates/` or a folder below that
    path like :file:`EXT:my_extension/Resources/Private/Templates/MyPages`.

    The templates are expected in a subfolder :path:`Pages` or
    :path:`pages`.

>     Fluid partials are searched for in a subfolder called :path:`Partials` or

Reviewer 2: ```suggestion
    Fluid partials are looked up in a sub-directory called :path:`Partials` or
```

Is :path: a mapped property? Thought I only used ":file:" always and need to memorize then :)

    :path:`partials`, layouts in :path:`Layouts` or :path:`layouts`.

    The name of the used page layout (:ref:`Backend layout <t3coreapi:be-layout>`)
    is resolved automatically.

>     The paths are considered from the largest to the smallest priority.

Reviewer 2: ```suggestion
    The paths are evaluated from highest to lowest priority.
```

..  _cobj-pageview-paths-example:

Example: Define a path that contains all templates
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> This is a basic definition of a `PAGEVIEW` object with only one path to the

Reviewer 2: ```suggestion
This is a basic definition of a :fluid:`PAGEVIEW` object with only one path to the
```

templates:

..  code-block:: typoscript
    :caption: EXT:my_sitepackage/Configuration/TypoScript/setup.typoscript

    page = PAGE
    page.10 = PAGEVIEW
    page.10.paths.100 = EXT:my_sitepackage/Resources/Private/Templates/

The content of the folder could look like this:

*   EXT:my_sitepackage/Resources/Private/Templates/

    *   Layouts

        *   Default.html
        *   WithoutHeader.html

    *   Pages

        *   Default.html
        *   StartPage.html
        *   TwoColumns.html
        *   With_sidebar.html

    *   Partials

        *   Footer.html
        *   Sidebar.html
        *   Menu.html

If a page has a backend layout named "with_sidebar", the template file is
then resolved

> to :file:`EXT:my_sitepackage/Resources/Private/Templates/Pages/With_sidebar.html`.

Reviewer 2: Oh, didn't catch that, thought it would resolve to CamelCase. Odd.

Author: It was mentioned in the changelog

If the backend layout is named "TwoColumns" it is resovled to
:file:`EXT:my_sitepackage/Resources/Private/Templates/Pages/TwoColumns.html`.

For all these templates
:ref:`partials <t3viewhelper:render_partial>`
are expected in folder
:path:`EXT:my_sitepackage/Resources/Private/Templates/Pages/Partials` and
:ref:`layouts <t3viewhelper:typo3fluid-fluid-layout>` in
:path:`EXT:my_sitepackage/Resources/Private/Templates/Pages/Layouts`.

..  _cobj-pageview-paths-example-extended:

Example: Define fallbacks for a template paths
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You can use the paths

Reviewer 2: I think we'll need to unify this PR to use either always "directory" OR "folder" OR "path", else readers may think we mean different things somtime. I know some people refer to "typoscript notation with a.config.key" as a "typoscript path".

(Personally I'd say let's go for "directory" at least on larger single pages, its beyond scope to unify this for all of the docs and hard to keep in synch with multiple contributees?)

Author: In this case it is even more confusing as the property is called "path"

Author: I tried rewording it
